app/src/main/java/com/babylon/wallet/android/presentation/transaction/analysis/TransactionAnalysisDelegate.kt

```diff
@@ -1,82 +1,146 @@
+    private val manifestSummaryToPreviewTypeAnalyser: ManifestSummaryToPreviewTypeAnalyser,
     private val cacheNewlyCreatedEntitiesUseCase: CacheNewlyCreatedEntitiesUseCase,
-    private val resolveNotaryAndSignersUseCase: ResolveNotaryAndSignersUseCase
+    private val sargonOsManager: SargonOsManager,
+    private val getProfileUseCase: GetProfileUseCase
 ) : DataHolderViewModelDelegate<TransactionReviewViewModel.Data, TransactionReviewViewModel.State>() {
 
     private val logger = Timber.tag("TransactionAnalysis")
 
-    suspend fun analyse() {
-        val manifestData = data.value.request.unvalidatedManifestData.also {
-            logger.v(it.instructions)
+    /**
+     * Runs analysis on each transaction type received. Resolves the preview type and returns the signers involved.
+     */
+    suspend fun analyse(): Result<Analysis> = when (data.value.request.transactionType) {
+        is TransactionType.PreAuthorized -> analysePreAuthTransaction(
+            manifestData = data.value.request.unvalidatedManifestData
+        )
+
+        else -> analyseTransaction(
+            manifestData = data.value.request.unvalidatedManifestData,
+            isInternal = data.value.request.isInternal
+        )
+    }.onSuccess { analysis ->
+        cacheNewlyCreatedResources(previewType = analysis.previewType)
+    }.onSuccess { analysis ->
+        _state.update {
+            it.copy(
+                isRawManifestVisible = analysis.previewType == PreviewType.NonConforming,
+                showRawTransactionWarning = analysis.previewType == PreviewType.NonConforming,
+                isLoading = false,
+                previewType = analysis.previewType,
+            )
         }
+    }.onFailure { error -> onError(error) }
 
-        transactionRepository.analyzeTransaction(
-            manifestData = manifestData,
-            isWalletTransaction = data.value.request.isInternal,
-            notaryPublicKey = data.value.ephemeralNotaryPrivateKey.toPublicKey()
-        ).then { transactionToReviewData ->
-            resolveNotaryAndSignersUseCase(
-                accountsAddressesRequiringAuth = transactionToReviewData.manifestSummary.addressesOfAccountsRequiringAuth,
-                personaAddressesRequiringAuth = transactionToReviewData.manifestSummary.addressesOfPersonasRequiringAuth,
-                notary = data.value.ephemeralNotaryPrivateKey
-            ).onSuccess { notaryAndSigners ->
-                data.update {
-                    it.copy(
-                        txToReviewData = transactionToReviewData,
-                        txNotaryAndSigners = notaryAndSigners
-                    )
-                }
-            }.map {
-                transactionToReviewData.transactionToReview.executionSummary
+    private suspend fun analyseTransaction(
+        manifestData: UnvalidatedManifestData,
+        isInternal: Boolean
+    ): Result<Analysis> = runCatching {
```

Comment:
So you are suggesting moving this code into `analyseTransactionPreview` and run it under the needed context?

Reply:
I'd place the analyse code that returns the model from Sargon inside the TransactionRepository and wrap it in the context with io dispatcher. If you agree I can do that as part of the analyse changes (error handling and the countdown).